### ISSUE TITLE
making casadi thread-safe

### ISSUE THREAD

**Issue by ghorn** [casadi/casadi#1355]
I'm opening this issue for discussion towards a thread-safe casadi.

There are use-cases which the `Parallelizer` class doesn't cover, such as running unit tests in parallel, or running callbacks in a separate thread to not block an iterative solve.

one first step is #1354

**Comment by jaeandersson** [comment]
My plan was to complete the separation of the temporary vector needed for evaluation and the Function datatype. Already now, for MXFunction there are only two work vectors for holding all the temporaries needed for evaluation. With that in place, multiple threads will be able to call the same Function instance at the same time with different temporary vectors. And cloning will become unnecessary. This also has other advantages, such as decreasing the overall overhead. And improved codegen.

**Comment by jaeandersson** [comment]
I'll close this issue and refer to #1356 for progress towards thread-safe evaluation.

**Comment by ghorn** [comment]
Let's close this `wontfix` and say that casadi supports thread-safe evaluation, but no guarantees on thread-safe symbolics

**Comment by jaeandersson** [comment]
Thread-safe symbolics would require a radically different syntax. You would have to carry around a kind of "world frame", which would contain all the static variables.
```python
import casadi
w = casad.world()
x = SX.sym(w, 'x')
c = SX.zeros(w, 3, 4)
```

**Comment by ghorn** [comment]
That's true, but it's pretty cumbersome. You could alternatively use a mutex for accessing shared mutable variables.

**Comment by jaeandersson** [comment]
Right. There are indeed ways to resolve this, maybe even without changing user syntax much. I guess just making a Function::init() a critical region would go a long way. I just don't think it's worth spending a lot of time on this at the moment, at least not for me.
